repose-aggregator/components/filters/keystone-v2/src/main/scala/org/openrepose/filters/keystonev2/KeystoneV2Filter.scala

```diff
@@ -0,0 +1,585 @@
+            case Success(_) => Pass
+            case Failure(e: MissingAuthTokenException) => Reject(SC_UNAUTHORIZED, Some(e.getMessage))
+            case Failure(e: InvalidTokenException) => Reject(SC_UNAUTHORIZED, Some(e.getMessage))
+            case Failure(e: InvalidTenantException) => Reject(SC_UNAUTHORIZED, Some(e.getMessage))
+            case Failure(e: UnparseableTenantException) => Reject(SC_UNAUTHORIZED, Some(e.getMessage))
+            case Failure(e: IdentityCommunicationException) => Reject(SC_BAD_GATEWAY, Some(e.getMessage))
+            case Failure(e: UnauthorizedEndpointException) => Reject(SC_FORBIDDEN, Some(e.getMessage))
+            case Failure(e: OverLimitException) =>
+              response.addHeader(HttpHeaders.RETRY_AFTER, e.retryAfter)
+              Reject(HttpServletResponse.SC_SERVICE_UNAVAILABLE, Some(e.getMessage))
+            case Failure(e) => Reject(SC_INTERNAL_SERVER_ERROR, Some(e.getMessage))
+          }
+        }
+
+      filterResult match {
+        case Pass =>
+          logger.trace("Processing completed, passing to next filter or service")
+          addIdentityStatusHeader(confirmed = true)
+          chain.doFilter(request, response)
+        case Reject(statusCode, message) =>
+          Option(config.getDelegating) match {
+            case Some(delegating) =>
+              logger.debug(s"Delegating with status $statusCode caused by: ${message.getOrElse("unspecified")}")
+
+              val delegationHeaders = buildDelegationHeaders(statusCode,
+                "keystone-v2",
+                message.getOrElse("Failure in the Keystone v2 filter").replace("\n", " "),
+                delegating.getQuality)
+
+              addIdentityStatusHeader(confirmed = false)
+              delegationHeaders foreach { case (key, values) =>
+                values foreach { value =>
+                  request.addHeader(key, value)
+                }
+              }
+
+              chain.doFilter(request, response)
+
+              logger.trace(s"Processing response with status code: $statusCode")
+
+              val wwwAuthenticateHeader = response.getHeader(CommonHttpHeader.WWW_AUTHENTICATE)
+
+              response.getStatus match {
+                case HttpServletResponse.SC_UNAUTHORIZED | HttpServletResponse.SC_FORBIDDEN =>
+                  if (DELEGATED.equalsIgnoreCase(wwwAuthenticateHeader.trim)) {
+                    logger.debug("The origin service could not authenticate the delegated request")
+                    response.setHeader(CommonHttpHeader.WWW_AUTHENTICATE, keystoneAuthenticateHeader)
+                  }
+                case HttpServletResponse.SC_NOT_IMPLEMENTED =>
+                  if (DELEGATED.equalsIgnoreCase(wwwAuthenticateHeader.trim)) {
+                    logger.error("Configured to delegate, but the origin service does not support delegation")
+                    response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR)
```

Comment:
Yeah, I went back and forth on it. I think I finally left it in based on some of what I read in Jorge's AuthN OpenStack Identity usage paper: https://wiki.openstack.org/wiki/Openstack-authn#Exchanging_User_Information

I think it's a safe thing to leave in, and it provides a little more information, but I have no problem removing it either. I'll wait on confirmation either way.